packages/zilliqa-js-subscriptions/src/ws.ts

```diff
@@ -0,0 +1,203 @@
+//  This file is part of Zilliqa-Javascript-Library.
+//
+//  This program is free software: you can redistribute it and/or modify
+//  it under the terms of the GNU General Public License as published by
+//  the Free Software Foundation, either version 3 of the License, or
+//  (at your option) any later version.
+//
+//   This program is distributed in the hope that it will be useful,
+//   but WITHOUT ANY WARRANTY; without even the implied warranty of
+//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+//   GNU General Public License for more details.
+//
+//   You should have received a copy of the GNU General Public License
+//   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+import mitt from 'mitt';
+import { w3cwebsocket as W3CWebsocket } from 'websocket';
+import {
+  MessageType,
+  NewBlockQuery,
+  NewEventQuery,
+  QueryParam,
+  SocketConnect,
+  SocketState,
+  StatusType,
+  SubscriptionOption,
+  Unsubscribe,
+} from './types';
+
+export class WebSocketProvider {
+  public static NewWebSocket(
+    url: string,
+    options?: SubscriptionOption,
+  ): WebSocket | W3CWebsocket {
+    // tslint:disable-next-line: no-string-literal
+    if (typeof window !== 'undefined' && (<any>window).WebSocket) {
+      // tslint:disable-next-line: no-string-literal
+      return new WebSocket(url, options !== undefined ? options.protocol : []);
+    } else {
+      const headers = options !== undefined ? options.headers || {} : undefined;
+      const urlObject = new URL(url);
+      if (
+        headers !== undefined &&
+        !headers.authorization &&
+        urlObject.username &&
+        urlObject.password
+      ) {
+        const authToken = Buffer.from(
+          `${urlObject.username}:${urlObject.password}`,
+        ).toString('base64');
+        headers.authorization = `Basic ${authToken}`;
+      }
+
+      return new W3CWebsocket(
+        url,
+        options !== undefined ? options.protocol : undefined,
+        undefined,
+        headers,
+        undefined,
+        options !== undefined ? options.clientConfig : undefined,
+      );
+    }
+  }
+
+  url: string;
+  emitter: mitt.Emitter;
+  handlers: any = {};
+  websocket: WebSocket | W3CWebsocket;
+  subscriptions: any;
+
+  // basically, options is a collection of metadata things like protocol or headers
+  constructor(url: string, options: any = {}) {
```

Comment:
Can we use `SubscriptionOption` instead of `any`?

Reply:
Yes. definitely.